runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -409,6 +428,45 @@ function LanguageTree:_add_injections()
   return query_time
 end
 
+--- Run an asynchronous parse, calling {on_parse} when complete.
+---
+--- @param range boolean|Range|nil
+--- @param on_parse fun(trees: TSTrees, timeout_expired: boolean)
+function LanguageTree:_async_parse(range, on_parse)
+  coroutine.resume(coroutine.create(function()
+    local co = coroutine.running()
+    local ct = self:_buf().changedtick
+    local total_parse_time = 0
+    local redrawtime = vim.o.redrawtime
+    ---@type TSTrees, boolean
+    local trees, unfinished
+
+    while total_parse_time <= redrawtime do
+      -- If buffer was changed in the middle of parsing, reset parse state
+      if self:_buf().changedtick ~= ct then
+        self._parser:reset()
+        self:invalidate()
+        total_parse_time = 0
+        coroutine.yield()
```

Comment:
This `yield` looks wrong to me since we don't `resume` before. I would have thought this works properly without the yields.

Reply:
Would it cause an infinite number of parses to be called if a user holds a key in a big file? 

The old parser doesn't finish under 3ms and goes to the next iteration, sees that the buffer has changed, resets the state and continues to parse. Meanwhile the new parser (called in `on_win()`) also starts to parse. And this happens every time a key is typed.

Reply:
Pulled in the version that does not use coroutines, so this yield isn't needed